### PR TITLE
Create input_schedule.js

### PR DESCRIPTION
**Pull request by Simran230**
Created a stub to add recurring schedule for service workers. The code describes a boolean which returns whether the schedule has been successfully inputed into the application or not.

### CODE REVIEW

**Comment by Simran230** [comment]
Scrum Board Ticket: S4 T2: Create a stub for proper DB availability schedule 
Story: Story 4 Adding same recurring schedule 
Description: Involves creating a public method with an agreed upon method name, input parameters, output format, which verifies that the weekly recurring schedule has been added to the DB.
Testing Criteria: Tested code by creating a Boolean to see whether the service worker has successfully inputted their weekly schedule or not.

File: server/models/handlers/input_schedule.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// Appointments occur at intervals of 30 minutes for each session for a 24 hour period Monday to Friday 
+// Service worker must input their avability at each corresponding school for the week. This recurres for 4 weeks
+//( int school_id_, int service_worker_id_, time[] start_times, string[] available, int[] days_available) 
+// Days Available - Monday = 1; Tuesday = 2; Wednesday = 3; Thursday = 4; Friday = 5
+//Output: Boolean Variable 
+async function inputSchedule(schoolID, workerID, slotID, status, daysAvailable) {
+    
+  var did_it_work = false;
+        //* insert (day, time, worker_id)
+ var did_it_work = true;
```
Comment (catai9) [inline server/models/handlers/input_schedule.js]: `var` is only needed for new variables. Since `did_it_work` was already declared in line 9, the `var` keyword can be omitted from line 13.

File: server/models/handlers/input_schedule.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Create stub to input weekly worker schedule into database 
+// Appointments occur at intervals of 30 minutes for each session for a 24 hour period Monday to Friday 
+// Service worker must input their avability at each corresponding school for the week. This recurres for 4 weeks
```
Comment (catai9) [inline server/models/handlers/input_schedule.js]: Code comments should be at a higher abstraction than the code. Thus, it might be better to omit line 4 or reword it to be more human readable.

**Review by catai9 — CHANGES_REQUESTED** [review]
Looks good, some minor changes required

File: server/models/handlers/input_schedule.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Create stub to input weekly worker schedule into database
```
Comment (catai9) [inline server/models/handlers/input_schedule.js]: Also make sure to implement the router so the front end knows what endpoint to call. E.g. `api/login`. Refer to routes folder, particularly public/private file for examples. Your endpoint will most likely be in the private file (since it's for a particular user).

**Comment by catai9** [comment]
Also, make sure to add automated unit tests in jest (refer to test folder for examples).

**Review by catai9 — DISMISSED** [review]
Please remove the .github/workflows files to prevent issues in deployment.

File: server/routes/private.js [hunk 4]
```diff
@@ -49,6 +50,41 @@ router.post('/api/book-appointment', async (req, res) => {
 
 });
 
+router.post('/api/recurring_schedule', async (req, res) => {
+    // Validate appropriate parameters are passed into the book appointment endpoint.
+    // Verification to make sure appointments are only made for future dates and that the worker is available during that time will be handled by the front end.
+    const paramSchema = Joi.object({
+
```
Comment (catai9) [inline server/routes/private.js]: Please remove unnecessary whitespace to increase code readability for the future.

File: server/routes/private.js [hunk 5]
```diff
@@ -49,6 +50,41 @@ router.post('/api/book-appointment', async (req, res) => {
 
 });
 
+router.post('/api/recurring_schedule', async (req, res) => {
```
Comment (catai9) [inline server/routes/private.js]: I would suggest changing the endpoint to add-recurring-schedule as per suggested coding naming conventions. Since this endpoint would change the database, the verb would signify that the endpoint manipulates the database.

File: server/routes/private.js [hunk 6]
```diff
@@ -49,6 +50,41 @@ router.post('/api/book-appointment', async (req, res) => {
...
+    const schoolId = query.schoolID ? query.schoolID : null;
+    const workerId = query.workerID ? query.workerID : null;
+    const status = query.status ? query.status : null;
+    const date = query.date ? query.date: null; 
+
+    const { error } = paramSchema.validate({workerTimeslotId, slotId, schoolId, workerId, status, date});
+
+    if (!_.isNil(error)) res.send(error);
+
+    // Attempts to insert the appointment into the database.
```
Comment (catai9) [inline server/routes/private.js]: I would adjust the comment to say add worker availability instead of appointment to avoid confusion.

File: server/test/handlers/inputSchedule.test.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+const appointmentHandler = require('../../models/handlers/workerTimeslot');
+
+describe('testing invalid user schedule input', () => {
+    beforeEach(() => {
+        jest.resetModules(); // Clears any cache between tests.
+    });
+
+    test('rejection of null values for worker and school ID', async () => {
+        // Arrange
+        const testworkerId = null;
```
Comment (catai9) [inline server/test/handlers/inputSchedule.test.js]: For greater consistency, I would suggest naming the variables with lowerCamelCase, i.e., testWorkerId.

File: server/test/handlers/inputSchedule.test.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        const testschoolID = 'Invalid School ID';
+
+        // Act
+        const user = await authenticateHandler.addWorkerTimeslot(testworkerID, testschoolID);
+        // Assert
+        expect(user).toBe(null);
+    });
+});
+
+describe('testing valid user validation', () => {
```
Comment (catai9) [inline server/test/handlers/inputSchedule.test.js]: Is this section purposeful or was it just accidentally kept from copying?

Comment (Simran230) [reply server/test/handlers/inputSchedule.test.js]: It was purposefully kept! Would you like me to remove it?

**Comment by catai9** [comment]
Hi Simran! Great improvements in code, there are some minor concerns and I would suggest updating your branch with the latest master changes as there are conflicts that need to be resolved.

File: server/test/handlers/inputSchedule.test.js (hunk 7 above)
Comment (Simran230) [reply server/test/handlers/inputSchedule.test.js]: Thank you @catai9 for your review! I will change the naming convention!

File: server/routes/private.js (hunk 6 above)
Comment (Simran230) [reply server/routes/private.js]: Thank you @catai9 I've added the change!

File: server/routes/private.js (hunk 5 above)
Comment (Simran230) [reply server/routes/private.js]: Thanks @catai9 I've modified the end point to add_recurring_schedule

File: server/routes/private.js (hunk 4 above)
Comment (Simran230) [reply server/routes/private.js]: Thanks @catai9 I've fixed the unnecessary whitespaces!

File: server/test/handlers/inputSchedule.test.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
+const appointmentHandler = require('../../models/handlers/workerTimeslot');
```
Comment (catai9) [inline server/test/handlers/inputSchedule.test.js]: The tests are not being run properly as the actual handler file is named `input_schedule` (it might be better to rename the handler file to `schedule` to increase flexibility as it may contain other methods, such as fetching schedules in the future). Remember to keep file names consistent and I would rename the `appointmentHandler` to the appropriate name used below, perhaps `scheduleHandler` to improve clarity of the specific handler.

File: server/test/handlers/inputSchedule.test.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    test('rejection of null values for worker and school ID', async () => {
+        // Arrange
+        const testworkerId = null;
+        const testschoolId = null;
+
+        // Act
+        const user = await workerTimeslotHandler.addWorkerTimeslot(testworkerID, testschoolID);
+
+        // Assert
+        expect(user).toBe(null);
```
Comment (catai9) [inline server/test/handlers/inputSchedule.test.js]: The current method is returning true instead of null for any and all inputted values so the tests would be failing.

File: server/test/handlers/inputSchedule.test.js (hunk 9 above)
Comment (Simran230) [reply server/test/handlers/inputSchedule.test.js]: Thanks @catai9 for your comment! I've updated and changed the handler name. Please let me know if this is what you're looking for!

File: server/test/handlers/inputSchedule.test.js (hunk 10 above)
Comment (Simran230) [reply server/test/handlers/inputSchedule.test.js]: @catai9 would you be able to clarify your comment here?

Comment (catai9) [reply server/test/handlers/inputSchedule.test.js]: You created a stub which currently returns true regardless of the parameters passed in. Thus, any automated tests would fail as you are asserting them to be equal to null but the method is in fact returning true. I would suggest commenting out these lines (so they can be kept and referred to later on) but writing a general test that just tests that your method currently does return the intended value of true. Refer to #17 for an example.

Another side note is that the addWorkerTimeslot method returns a boolean value rather than an object so in line 14 (and other associated lines), I would recommend naming the variable to be isSuccessfullyInserted or another similar adjective to prevent confusion.

File: .github/workflows/src/user.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+export default (sequelize, Datatypes) => {
```
Comment (catai9) [inline .github/workflows/src/user.js]: Please remove all `.github/workflows` files that you have added, `availability.js`, `token.js`, `user.js` to avoid deployment issues as they are not required for your code to run.

File: server/routes/private.js [hunk 12]
```diff
@@ -49,6 +50,37 @@ router.post('/api/book-appointment', async (req, res) => {
 
 });
 
+router.post('/api/add_recurring_schedule', async (req, res) => {
+    // Validate appropriate parameters are passed into the book appointment endpoint.
+    // Verification to make sure appointments are only made for future dates and that the worker is available during that time will be handled by the front end.
+    const paramSchema = Joi.object({
+        workerTimeslotId: Joi.number().integer().required(),
```
Comment (catai9) [inline server/routes/private.js]: The `workerTimeslotId` would be automatically incremented. Thus, I would remove it from the parameter scheme as it does not need to be passed into the endpoint.

File: server/routes/private.js [hunk 13]
```diff
@@ -49,6 +50,37 @@ router.post('/api/book-appointment', async (req, res) => {
...
+        slotId: Joi.number().integer().required(),
+        schoolId: Joi.number().integer().required(),
+        workerId: Joi.number().integer().required(), 
+        status:Joi.string().max(300).required(),
```
Comment (catai9) [inline server/routes/private.js]: I would recommend making the status optional and set it to have a default value of available to increase flexibility of this endpoint to future uses.

File: server/routes/private.js [hunk 14]
```diff
@@ -49,6 +50,37 @@ router.post('/api/book-appointment', async (req, res) => {
...
+        date: Joi.string().allow(null).max(300)
```
Comment (catai9) [inline server/routes/private.js]: Should be of type date rather than string, i.e., use `Joi.date()`. Refer to #17 for an example use of Joi.date.

File: server/routes/private.js (hunk 12 above)
Comment (Simran230) [reply server/routes/private.js]: Thanks @catai9 workerTimeslotID is removed!

**Comment by catai9** [comment]
I ran the automated tests and one of the tests had failed. Please update accordingly.
image

**Comment by Simran230** [comment]
> I ran the automated tests and one of the tests had failed. Please update accordingly.
> image

Hi Amy! Thank you for testing the code! What does this error mean?

**Comment by prdbondi** [comment]
Ok either there is no handler and I'm the one who made it, or the naming is wrong and there is another handler file called timeslot or something. Not sure how to reconcile if there is no handler on simran branch other than i give her my code for the handler.

File: server/test/handlers/inputSchedule.test.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
+const workerTimeslotHandler = require('../../models/handlers/workerTimeslot');
```
Comment (catai9) [inline server/test/handlers/inputSchedule.test.js]: Please note that the handler file is called `timeslot` rather than `workerTimeslot` so I would suggest adjusting the file name here to fix the failing tests.

Comment (Simran230) [reply server/test/handlers/inputSchedule.test.js]: Thank you @catai9 for your comment! I've made the changes accordingly.

File: server/routes/private.js [hunk 16]
```diff
@@ -50,6 +51,34 @@ router.post('/api/book-appointment', async (req, res) => {
...
 });
 
+router.post('/api/add-recurring-schedule', async (req, res) => {
+    // Validate appropriate parameters are passed into add recurring schedule. 
+    const paramSchema = Joi.object({
+        slotId: Joi.number().integer().required(),
+        schoolId: Joi.number().integer().required(),
+        workerId: Joi.number().integer().required(),
+        status: Joi.string().max(30), // The default value is available which is handled by the database Create Table logic. 
+        date: Joi.date().iso(),
```
Comment (catai9) [inline server/routes/private.js]: What happens if no date is specified? Does it input this schedule for all weekdays? I would suggest adding in a comment to increase shared understanding of the functionality of this endpoint.

File: server/test/handlers/inputSchedule.test.js (hunk 15 above)
Comment (catai9) [reply server/test/handlers/inputSchedule.test.js]: Did you remember to push? I still see the previous code referring to `workerTimeslot`.

File: server/routes/private.js (hunk 16 above)
Comment (Simran230) [reply server/routes/private.js]: @catai9 that was my envision. Is there anything you have in mind? :)

Comment (Simran230) [reply server/routes/private.js]: Perfect I will add in a comment! :) 
Thank you Amy!

File: server/routes/private.js [hunk 17]
```diff
@@ -50,6 +51,34 @@ router.post('/api/book-appointment', async (req, res) => {
 
 });
 
+router.post('/api/add-recurring-schedule', async (req, res) => {
+    // Validate appropriate parameters are passed into add recurring schedule. 
+    const paramSchema = Joi.object({
+        slotId: Joi.number().integer().required(), //The worker must specify the slot(s) they are available
```
Comment (catai9) [inline server/routes/private.js]: They are only specifying one slotId here based on the Joi.number().integer() rather than an array so I would recommend removing (s) in the comment to avoid future confusion.

File: server/test/handlers/inputSchedule.test.js [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+//         expect(user).toBe(null);
+//     });
+// });
+
+// describe('testing valid user validation', () => {
+//     beforeEach(() => {
+//         jest.resetModules(); // Clears any cache between tests.
+//     });
+// });
+
```
Comment (Kyle199910) [inline server/test/handlers/inputSchedule.test.js:101]: was this test code, if so, comment at top that this is test code

File: server/test/handlers/inputSchedule.test.js (hunk 8 above)
Comment (catai9) [reply server/test/handlers/inputSchedule.test.js]: It should be commented out as the automated tests would otherwise fail and the deployment may encounter problems since there is no actual test within the describe block.

File: server/routes/private.js (hunk 16 above)
Comment (catai9) [reply server/routes/private.js]: It may be better to have a start and an end date to allow for more flexibility in the future. However, for the time being, this is sufficient to allow the user to enter their weekly recurring schedule into the system.

File: server/test/handlers/inputSchedule.test.js (hunk 18 above)
Comment (Simran230) [reply server/test/handlers/inputSchedule.test.js:101]: @Kyle199910 this test could have been used for Peter's test

Comment (catai9) [reply server/test/handlers/inputSchedule.test.js:101]: Right, excellent point, I also agree with Kyle's concern. I would recommend adding a brief comment at the top signifying that this is test code that can be leveraged once the logic is fully implemented for this functionality.

File: server/routes/private.js (hunk 17 above)
Comment (Simran230) [reply server/routes/private.js]: Thanks @catai9 completed!

File: server/test/handlers/inputSchedule.test.js (hunk 18 above)
Comment (Simran230) [reply server/test/handlers/inputSchedule.test.js:101]: @Kyle199910 @catai9 thank you for your response! A comment has been made!

**Review by Kyle199910 — APPROVED** [review]
Verified that code passes npm test: Code comments mentioned were resolved.
